fix(uninstall): Correct package uninstallation with dotted attr_path

tanjadev opened the pull request:
This resolves the issue causing confusing messages when uninstalling packages with a dotted attr_path (e.g., foo.bar). The package name is now correctly parsed, allowing such packages to be uninstalled from the environment.

fixes #1714

ghudgins reviewed (approved):
functionality LGTM. (not a code review)

zmitchell commented on cli/flox-rust-sdk/src/models/manifest.rs:
```diff
@@ -1228,11 +1228,17 @@ pub fn remove_packages(
-        if !installs_table.contains_key(pkg) {
+
+        // Get package id from the nested key or keep the package name as the id
+        let pkg_id = CatalogPackage::from_str(pkg)
```
I don't think this logic is quite right. What we want is (in this order):
```
if str == an existing install_id
    use that id
else if str == the pkg-path for an existing pkg
    use the id for that pkg
else
    error
```
What we have here is:
```
if str can be parsed as a pkg
    use the inferred id of that pkg
else
    use str as the id
```

Without checking (not at a computer at the moment) I think this would miss `flox uninstall cargo` for `mycargo.pkg-path = "cargo"` since it would parse `cargo` into an inferred install id of `cargo` rather than looking up "the id of the package whose pkg-path is `cargo`".

zmitchell replied:
It would be good to have a test case for that scenario. The ticket is asking us to support uninstalling via pkg-path, not just an install id that has a dot in it.

tanjadev replied:
Yep, that was due to my initial misinterpretation of the ACs 🤦‍♀️ Now the logic should match what you've described. I've added some unit tests that cover different cases (as discussed with @ysndr).

brendaneamon commented:
> With the following chunk in the manifest:
> 
> ```
> tv1 = {pkg-path = "python311Packages.torchvision-bin", systems=["aarch64-darwin", "x86_64-darwin"],  pkg-group="darwin"}
> tv2 = {pkg-path = "python311Packages.torchvision-bin", systems=["aarch64-linux", "x86_64-linux"],  pkg-group="linux"}
> ```
> 
> I get a failure for `flox uninstall python311Packages.torchvision-bin`. This is the case I had in mind for the `XYZ is ambiguous, please use the install-id, one of [ tv1, tv2 ]`. Or something like that.

@billlevine, @tanjadev, if I'm not mistaken, this is related to the following issue: https://github.com/flox/flox/issues/1512

billlevine commented on cli/flox-rust-sdk/src/models/manifest.rs:
```diff
@@ -897,8 +897,7 @@ pub enum TomlEditError {
     MissingInstallTable,
-    /// Tried to uninstall a package that wasn't installed
-    #[error("couldn't uninstall '{0}', wasn't previously installed")]
+    #[error("couldn't find package '{0}' in the manifest")]
```
nb: Since the _item_ to remove could be a package or an install id, would it be better to simplify this message as `couldn't find '{0}' in the manifest` ?

zmitchell replied:
I'd say yes

tanjadev replied:
Good point, the term "package" can be a bit ambiguous here.

In fact, now it's back to only being the install id, as we extract them in the Core Environment before calling the remove function. Initially, it was designed to handle both package and install id, but after feedback, I moved the handling to the Core Environment. I've updated the wording in the code and logs to "install id" to make it clearer.

tanjadev commented:
This one is ready for another round of review. I've changed a bit of wording and added some unit tests covering the following cases: 

- Return the install ID if it matches the user input.
- Return the install ID if a pkg-path matches the user input.
- Ensure that the install ID takes precedence over pkg-path when both are present, resulting in the deletion of the corresponding package.
- Throw an error if multiple pkg-paths match the user input.
- Throw an error if no install ID or pkg-path matches the user input.

zmitchell commented:
I'll give a look after the service management meeting 👍

zmitchell commented on cli/flox-rust-sdk/src/models/environment/core_environment.rs:
```diff
@@ -420,6 +424,45 @@ impl CoreEnvironment<ReadOnly> {
     }
 
+    fn get_install_ids_to_uninstall(
+        manifest: &TypedManifest,
```
Note to self: this can take a `&TypedManifestCatalog` instead since we're about to no longer allow locking v0  (`pkgdb`) manifests, which removes the need for `if let ...`

zmitchell commented on cli/flox-rust-sdk/src/models/environment/core_environment.rs:
```diff
@@ -1196,6 +1239,14 @@ pub enum CoreEnvironmentError {
+    PackageNotFound(String),
+    // Multiple packages match user input, must specify install_id
+    #[error(
+        "multiple packages match '{0}', please specify an install id from possible matches: {1:?}"
```
Note to self: May want to tweak how the matches are displayed by doing a `.join(", ")` into a `String` instead
